apps/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts

```diff
@@ -419,100 +419,44 @@ export class WorkspaceInstallManager extends BaseInstallManager {
-      '    ? clientPnpmfile.hooks.afterAllResolved(lockfile, context)',
-      '    : lockfile;',
-      '}',
-
-      // Set the preferred versions in the package, then call the original pnpmfile (if it exists)
-      'function readPackage(pkg, context) {',
-      '  setPreferredVersions(pkg.dependencies);',
-      '  setPreferredVersions(pkg.devDependencies);',
-      '  setPreferredVersions(pkg.optionalDependencies);',
-      '  return (clientPnpmfile && clientPnpmfile.hooks && clientPnpmfile.hooks.readPackage)',
-      '    ? clientPnpmfile.hooks.readPackage(pkg, context)',
-      '    : pkg;',
-      '}',
-
-      // Set the preferred versions on the dependency map. If the version on the map is an allowedAlternativeVersion
-      // then skip it. Otherwise, check to ensure that the common version is a subset of the specified version. If
-      // it is, then replace the specified version with the preferredVersion
-      'function setPreferredVersions(dependencies) {',
-      '  for (const name of Object.keys(dependencies || {})) {',
-      '    if (allPreferredVersions.hasOwnProperty(name)) {',
-      '      const preferredVersion = allPreferredVersions[name];',
-      '      const version = dependencies[name];',
-      '      if (allowedAlternativeVersions.hasOwnProperty(name)) {',
-      '        const allowedAlternatives = allowedAlternativeVersions[name];',
-      '        if (allowedAlternatives && allowedAlternatives.indexOf(version) > -1) {',
-      '          continue;',
-      '        }',
-      '      }',
-      '      let isValidRange = false;',
-      '      try {',
-      '        isValidRange = !!semver.validRange(preferredVersion) && !!semver.validRange(version);',
-      '      } catch {',
-      '      }',
-      '      if (isValidRange && semver.subset(preferredVersion, version)) {',
-      '        dependencies[name] = preferredVersion;',
-      '      }',
-      '    }',
-      '  }',
-      '}'
-    ];
-
     // Attempt to move the existing pnpmfile if there is one
     try {
       const pnpmfileDir: string = path.dirname(filename);
       await FileSystem.moveAsync({
         sourcePath: filename,
-        destinationPath: path.join(pnpmfileDir, clientPnpmfileName)
+        destinationPath: path.join(pnpmfileDir, 'clientPnpmfile.js')
       });
     } catch (error) {
       if (!FileSystem.isNotExistError(error)) {
         throw error;
       }
     }
 
-    // Write the shim pnpmfile to the original file path
-    await FileSystem.writeFileAsync(filename, pnpmfileContent.join(NewlineKind.Lf), {
-      ensureFolderExists: true
-    });
+    // Read the shim pnpmfile and replace the placeholders
+    let pnpmfileContent: string = await FileSystem.readFileAsync(
+      path.resolve(__dirname, '..', 'pnpm', 'PnpmfileShim.js')
+    );
+    pnpmfileContent = pnpmfileContent.replace(
+      /__semverPath/,
+      JSON.stringify(require.resolve('semver')).replace(/^"|"$/g, '')
```

Comment:
Would `.trimStart('"').trimEnd('"') be cleaner?

Reply:
Far as I can tell, that's not actually a thing. That was my first thought too

Reply:
This problem of *"How do I correctly escape strings pasted into the middle of source code?"* has numerous edge cases.  (What if Prettier rewrites your quotes? What if a developer doesn't understand exactly what context `__semverPath` should be used in? etc)

The problem will disappear if we instead use a conventional argument-passing mechanism such as reading the data from a separate JSON file.  We could also emit a JSON literal at the top of the file (like with HTML scripts), but self-modifying code makes me uncomfortable generally.  And it's harder for casual users to understand and debug.

Reply:
Done!